### ISSUE TITLE
Duplicate logrotate commands on Linux cause error

### ISSUE THREAD

**Issue by mcbsys** [Microsoft/OMS-Agent-for-Linux#926]
Originally posted as https://github.com/Microsoft/omi/issues/627 but was asked to post here instead.

It looks like log rotation was set up here https://github.com/Microsoft/omi/issues/531, so hopefully this is the right place to report an issue with that.

Yesterday, I created a new Ubuntu 18.04 VM from Microsoft image. I set "OS guest diagnostics" = On. This installed LinuxDiagnostic extension 3.0.119, which created /etc/logrotate.d/omsagent-LAD with this content:
```
/var/opt/microsoft/omsagent/LAD/log/omsagent.log {
    rotate 10 
    missingok
    notifempty
    compress
    size 100M
    copytruncate
}

/var/opt/microsoft/omsagent/log/urp.log {
    rotate 3
    missingok
    notifempty
    compress
    size 50k
    copytruncate
}
```
I also have Security Center - Security policy > Settings - Data Collection - Auto Provisioning set to On. So eventually OmsAgentForLinux 1.9.1 was auto-deployed. This created /etc/logrotate.d/omsagent-aaaaa1b3-1111-2222-bf39-700df40cffff with this content:
```
/var/opt/microsoft/omsagent/omsagent-aaaaa1b3-1111-2222-bf39-700df40cffff/log/omsagent.log {
    rotate 10 
    missingok
    notifempty
    compress
    size 100M
    copytruncate
}

/var/opt/microsoft/omsagent/log/urp.log {
    rotate 3
    missingok
    notifempty
    compress
    size 50k
    copytruncate
}
```
Because there are duplicate logrotate configurations on the same log file (urp.log), every day I get this via email:

> Cron <root@MYSERVER> test -x /usr/sbin/anacron || ( cd / && run-parts --report /etc/cron.daily )
> /etc/cron.daily/logrotate:
> error: omsagent-LAD:10 duplicate log entry for /var/opt/microsoft/omsagent/log/urp.log
> 

I'm not sure how these two versions of the OMS agent are supposed to interact, but somehow the log rotation needs to be consolidated. Maybe just a separate /etc/logrotate.d/urp file for urp.log so if it already exists, it doesn't create another one?

**Comment by dstrzelec** [comment]
Is there any news on this issue? I am experiencing the same thing.

**Comment by hestolz** [comment]
This will be fixed in an upcoming release, see #940 
ETA for the release including this change is July.

**Comment by maartenb** [comment]
It seems two log-rotate processes could end up working on the same logfile at the same time IIUC?
Why are two services logging to the same file? why is a single entry in /etc/logrotate.d not sufficient?

**Comment by mcbsys** [comment]
More like when two flavors of the OMS agent get installed, both use urp.log and and rotation instructions. I don't think the issue is that they rotate at the same time; rather, LInux is complaining that there are two potentially conflicting log rotation commands affecting the same file.

**Comment by maartenb** [comment]
Yes, and the proposed fix seems to be to hide that from logrotate ("Linux") rather then solve the problem. A better solution would be for each installed flavor log to its own log-file or provide a single configuration to logrotate.

**Comment by mikini** [comment]
I'm seeing a similar situation for an Ubuntu 16.04 VM.

There seems to be some confusion regarding the placement of osmagent logs on the system (assuming it is intentional, or a separate issue, that two osmagents are present in the first place). The situation can be summarised as (details below):

* Two logrotate configuration files are present referring to two different osmagent.log files but the same urp.log file which makes logrotate bark.
* The directory in which the urp.log file is supposed to be placed is however symlinked to the directory of one of the osmagent.log files (maybe meant as a "default osm instance" when reading logs?).
* Nowhere is an actual urp.log file present.

I can't help to suspect that maybe the urp.log file path in the logrotate config is just missing the "instance specific" path component (in the case below "LAD" and "54681bff-87c6-4913-b889-6d27245fb6e7" respectively.

**Details**
```
$ sudo ls -l /etc/logrotate.d/omsagent-*
-rw-r--r-- 1 root root 304 Jul 27 03:34 /etc/logrotate.d/omsagent-54681bff-87c6-4913-b889-6d27245fb6e7
-rw-r--r-- 1 root root 271 Sep 27 09:23 /etc/logrotate.d/omsagent-LAD
$ sudo cat /etc/logrotate.d/omsagent-*
/var/opt/microsoft/omsagent/54681bff-87c6-4913-b889-6d27245fb6e7/log/omsagent.log {
    rotate 10 
    missingok
    notifempty
    compress
    size 100M
    copytruncate
}

/var/opt/microsoft/omsagent/log/urp.log {
    rotate 3
    missingok
    notifempty
    compress
    size 50k
    copytruncate
}
/var/opt/microsoft/omsagent/LAD/log/omsagent.log {
    rotate 10 
    missingok
    notifempty
    compress
    size 100M
    copytruncate
}

/var/opt/microsoft/omsagent/log/urp.log {
    rotate 3
    missingok
    notifempty
    compress
    size 50k
    copytruncate
}

$ sudo realpath /var/opt/microsoft/omsagent/log/urp.log
/var/opt/microsoft/omsagent/54681bff-87c6-4913-b889-6d27245fb6e7/log/urp.log

$ sudo ls -l /var/opt/microsoft/omsagent/{54681bff-87c6-4913-b889-6d27245fb6e7,LAD}/log
/var/opt/microsoft/omsagent/54681bff-87c6-4913-b889-6d27245fb6e7/log:
total 12600
-rw-rw-rw- 1 omsagent omiusers       80 Nov  4 13:50 ODSIngestion.status
-rw-r--r-- 1 omsagent omiusers 12886117 Nov  4 13:49 omsagent.log
-rw-r----- 1 omsagent omiusers       74 Nov  4 02:21 urp.log.pos

/var/opt/microsoft/omsagent/LAD/log:
total 8
-rw-r--r-- 1 omsagent omiusers 7763 Oct 28 08:30 omsagent.log
$ 
```

**Comment by hestolz** [comment]
Thank you for your submission. As [previously announced](https://azure.microsoft.com/en-us/updates/were-retiring-the-log-analytics-agent-in-azure-monitor-on-31-august-2024/), the Log Analytics agent has been **deprecatated** and has no support as of **August 31, 2024.** If you use the Log Analytics agent to ingest data to Azure Monitor, [migrate now to the new Azure Monitor agent](https://docs.microsoft.com/en-us/azure/azure-monitor/agents/azure-monitor-agent-migration). As part of repo archival, open issues and pull requests will be closed.
